checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py

```diff
@@ -0,0 +1,293 @@
+            Fragment(
+                index=numpy_utils.resolve_slice(index, self.global_shape),
+                value=data,
+            )
+            for index, data in self.replica_slices
+        ],
+    )
+    if fragments.fragments:
+      validate_fragments_can_be_stacked(fragments)
+    if not fragments.is_degenerate():
+      assert self.local_shape == fragments.fragments[0].shape
+    return fragments
+
+
+def _get_replica_counts(arr: jax.Array) -> Callable[[jax.Shard], int]:
+  """Produces a mapping from addressable shards to global replication count."""
+
+  # slice objects are not hashable before python 3.12.
+  def hashable_slices(slices: list[slice]):
+    return tuple(((s.start, s.stop, s.step) for s in slices))
+
+  counts = defaultdict(int)
+  for index in arr.sharding.devices_indices_map(arr.shape).values():
+    counts[hashable_slices(index)] += 1
+
+  return lambda shard: counts[hashable_slices(shard.index)]
+
+
+def get_replica_slices(
+    arr: jax.Array,
+    replica_id: Optional[int],
+    use_replica_parallel: bool,
+) -> ReplicaSlices:
+  """Returns the replica slices a given replica is responsible for.
+  Does not transfer allocate or transfer any data."""
+  Result = tuple[list[ReplicaSliceOnDevice], Shape]
+  shard0 = arr.addressable_shards[0]
+
+  # single-replica: a single replica saves an entire shard.
+  def pick_single_replica() -> Result:
+    # Omitting the replica id just picks the first addressable shard's replica
+    # id so that the process writes each of its addressable shards exactly
+    # once. (This is the desired behavior for local checkpointing.)
+    target_replica_id = replica_id or shard0.replica_id
+    rslices = [
+        ReplicaSliceOnDevice(
+            replica_id=shard.replica_id,
+            index=shard.index,
+            unsliced_data=shard.data,
+            slice_args=None,
+        )
+        for shard in arr.addressable_shards
+        if shard.replica_id == target_replica_id
+    ]
+    local_shape = shard0.data.shape
+    return rslices, local_shape
+
+  # replica-parallel: every replica saves part of a shard.
+  # Logic based on axlearn: https://github.com/apple/axlearn/blob/226d27ab7569668f2c38a35cf32d5dc5190ebdbb/axlearn/common/array_serialization.py#L75
+  def maybe_pick_replica_parallel() -> Optional[Result]:
+    assert (
+        replica_id is not None
+    ), 'use_replica_parallel is incompatible with local checkpointing'
+
+    # Check whether replica-parallel applies: we are dealing with non-empty
+    # shards, we have more than one replica, and some dimension of the shards
+    # is evenly divisible across replicas.
+    if math.prod(shard0.data.shape) == 0:
+      return None
+    replica_counts = _get_replica_counts(arr)
+    replica_count = replica_counts(shard0)
+    try:
+      axis = next(
+          axis_index
+          for axis_index, axis_size in enumerate(shard0.data.shape)
+          if replica_count > 1 and axis_size % replica_count == 0
```

Comment:
It seems fairly easy to construct examples for which this approach does not work. e.g. let's say there are 8 devices, global array of (11, 4, 7), where the middle axis is sharded 4 ways, giving 2 replicas per shard, shape (11, 1, 7). replica_count does not evenly divide any dimension so we would have to fall back on the old approach.

I guess we are just assuming that realistic use cases will not use weird dimensions like this.

Reply:
Indeed, that's a limitation of the current approach.

I haven't looked into it too deeply, but in principle it seems like we just allow the last replica of each shard to write a slice that's not exactly of `local_shape` but shorter along the chosen axis. We'd then choose the slicing axis in a way that prefers an evenly-divisible axis, but may fall back to a non-evenly-divisible one.

We'd need to check how much that degrades tensorstore write/read performance. (Perhaps you've experimented with that before in the context of fragments?)

Reply:
We've never experimented with uneven fragments when writing, so I also don't know about what the performance impacts would be, if any. I agree this can be a follow-up. Maybe worth leaving a small TODO with a note about this aspect.